SDL/Quintuplet.cu

```diff
@@ -602,138 +635,394 @@ __device__ bool SDL::passT5RZConstraint(struct SDL::modules& modulesInGPU, struc
+    }
+    else if((y3-y_center)<0 && (y1-y_center)<0) 
+    {
+        if (slope3c>slope1c) charge=-1; 
+        else if (slope3c<slope1c) charge=1;
+        if (slope1c<0 && slope3c>0) charge=1;
+        if (slope1c>0 && slope3c<0) charge=-1;
+    }
+    else if ((y3-y_center)<0 && (y1-y_center)>0)
+    {
+        if ((x3-x_center)>0 && (x1-x_center)>0) charge = 1;
+        else if ((x3-x_center)<0 && (x1-x_center)<0) charge = -1;
+    }
+    else if ((y3-y_center)>0 && (y1-y_center)<0)
     {
-        slope = (z3 - z1)/(rt3 - rt1);
+        if ((x3-x_center)>0 && (x1-x_center)>0) charge = -1;
+        else if ((x3-x_center)<0 && (x1-x_center)<0) charge = 1;
     }
-    float residual4 = (layer4 <= 6)? ((z4 - z1) - slope * (rt4 - rt1)) : ((rt4 - rt1) - (z4 - z1)/slope);
-    float residual5 = (layer4 <= 6) ? ((z5 - z1) - slope * (rt5 - rt1)) : ((rt5 - rt1) - (z5 - z1)/slope);
 
-    // creating a chi squared type quantity
-    // 0-> PS, 1->2S
-    residual4 = (moduleLayer4 == 0) ? residual4/2.4f : residual4/5.0f;
-    residual5 = (moduleLayer5 == 0) ? residual5/2.4f : residual5/5.0f;
+    float pseudo_phi = atan((y_init-y_center)/(x_init-x_center)); //actually represent pi/2-phi, wrt helix axis z
+    float Pt=inner_pt, Px=Pt*abs(sin(pseudo_phi)), Py=Pt*abs(cos(pseudo_phi));
 
-    const float RMSE = sqrtf(0.5 * (residual4 * residual4 + residual5 * residual5));
+    if (x_init>x_center && y_init>y_center) //1st quad
+    {
+        if (charge==1) Py=-Py;
+        if (charge==-1) Px=-Px;
+    }
+    if (x_init<x_center && y_init>y_center) //2nd quad
+    {
+        if (charge==-1) {Px=-Px; Py=-Py;}
+    }
+    if (x_init<x_center && y_init<y_center) //3rd quad
+    {
+        if (charge==1) Px=-Px;
+        if (charge==-1) Py=-Py;
+    }        
+    if (x_init>x_center && y_init<y_center) //4th quad
+    {
+        if (charge==1) {Px=-Px; Py=-Py;}
+    }
+
+    if (moduleType3==0){
+        if (x4<x3 && x3<x2) Px=-abs(Px);
+        if (x4>x3 && x3>x2) Px=abs(Px);
+        if (y4<y3 && y3<y2) Py=-abs(Py);
+        if (y4>y3 && y3>y2) Py=abs(Py);
+    }
+    else if(moduleType3==1)
+    {
+        if (x3<x2 && x2<x1) Px=-abs(Px);
+        if (x3>x2 && x2>x1) Px=abs(Px);
+        if (y3<y2 && y2<y1) Py=-abs(Py);
+        if (y3>y2 && y2>y1) Py=abs(Py);        
+    }
+
+    //to get Pz, we use pt/pz=ds/dz, ds is the arclength between MD1 and MD3.
+    float AO=sqrt((x1-x_center)*(x1-x_center)+(y1-y_center)*(y1-y_center));
+    float BO=sqrt((x_init-x_center)*(x_init-x_center)+(y_init-y_center)*(y_init-y_center));
+    float AB=sqrt((x1-x_init)*(x1-x_init)+(y1-y_init)*(y1-y_init)); 
+    float dPhi = acos((AO*AO+BO*BO-AB*AB)/(2*AO*BO));
+    float ds=innerRadius/100*dPhi;
+
+//    float ds = sqrt((y_init-y1)*(y_init-y1)+(x_init-x1)*(x_init-x1)); //large ds->smallerPz->smaller residual
+    float Pz=(z_init-z1)/ds*Pt;
+    float p = sqrt(Px*Px+Py*Py+Pz*Pz);
+
+    float B = 3.8112;
+    float a = -0.299792*B*charge;
```

Comment:
These look like constants. Should we define them in the proper file (`Constants.cu`)?

Reply:
Yes, it's a good suggestion to put constants in the Constants.cu file. But I think here I only use it in this certain function, so I guess we can just use a simple number here? If we later use this number somewhere else in the code, we can surely put into that cu file

Reply:
I think that it is good practice to keep all of the constants in one file and load them from there - the opposite cases means more definitions, which leads to longer and more bug-prone code. Could you please modify so that these are moved to the appropriate file with the constants (if they are not already there) and are loaded from there?

Reply:
I put the magnetic field B into constant.cu . The -0.299792 is a parameter in helix formula and it's not a physical const, so I leave it here